### PR TITLE
Fix NotebookInstanceV2 and verify against realGCP

### LINKED ISSUE #7168: Vertex AI Workbench KCC Support
### Checklist

- [x] I did not find a related open enhancement request.
- [x] I understand that enhancement requests filed in the GitHub repository are by default low priority.
- [x] If this request is time-sensitive, I have submitted a corresponding issue with [GCP support](https://cloud.google.com/support-hub).

### Describe the feature or resource

A large FSI customer is migrating their notebook workloads to Vertex AI Workbench and requires the service to be supported as a resource in Config Connector (KCC)

### Additional information

_No response_

### Importance

Impact: KCC support is a prerequisite for the customer to move their notebook environment into production. Without this support, they are unable to use their standard IaC processes to provision and manage Workbench instances, which is a blocker for their production go-live.

Urgency: This is a blocker. Customer needs to onboard the service by April/26

### PR DESCRIPTION
**Pull request by acpana**
Changes and tweaks for a real run for NotebookInstanceV2

Fixes #7168

```release-note
Fixes and improvements for NotebookInstanceV2 resource.
```

### CODE REVIEW

File: pkg/controller/direct/notebooks/notebookinstancev2_controller.go [hunk 1]
```diff
@@ -224,126 +232,130 @@ func (a *InstanceV2Adapter) Update(ctx context.Context, updateOp *directbase.Upd
...
-				desVM.Project = actVM.Project
-			}
-			if desVM.GetFamily() == "" && actVM.GetFamily() != "" {
-				desVM.Image = &notebookspb.VmImage_Family{Family: actVM.GetFamily()}
+
+		// Handle GceSetup (Infrastructure) fields
+		actGce := act.GetGceSetup()
+		desGce := des.GetGceSetup()
+
+		// Align immutable fields under GceSetup from actual to desired.
```
Comment (maqiuyujoyce) [inline pkg/controller/direct/notebooks/notebookinstancev2_controller.go:267]: It's out of scope, but I think it'll be helpful if we make this a util function: https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/pkg/controller/direct/composer/composerenvironment_defaults.go#L58

File: pkg/controller/direct/notebooks/notebookinstancev2_controller.go [hunk 2]
```diff
@@ -224,126 +232,130 @@ func (a *InstanceV2Adapter) Update(ctx context.Context, updateOp *directbase.Upd
...
-			desGCE.EnableIpForwarding = actGCE.EnableIpForwarding
-		}
-		// Copy/align ShieldedInstanceConfig defaults
-		if actGCE.ShieldedInstanceConfig == nil {
-			actGCE.ShieldedInstanceConfig = &notebookspb.ShieldedInstanceConfig{
-				EnableSecureBoot:          false,
-				EnableVtpm:                true,
-				EnableIntegrityMonitoring: true,
+
+			// Ensure all desired metadata exist in actual; otherwise it is changed
```
Comment (maqiuyujoyce) [inline pkg/controller/direct/notebooks/notebookinstancev2_controller.go:299]: This cannot handle the edge case that users remove entries from metadata.

I'm okay leaving this an open question in this PR. But we probably need a solution eventually.

File: pkg/test/resourcefixture/testdata/basic/notebooks/v1alpha1/notebookinstancev2/notebookinstancev2-container/_http_mock.log [hunk 3]
```diff
@@ -102,28 +159,215 @@ OK
...
+      "dataDisks": [
+        {
+          "diskEncryption": "GMEK",
+          "diskSizeGb": "100",
+          "diskType": "PD_STANDARD"
+        }
+      ],
+      "disablePublicIp": true,
+      "machineType": "n1-standard-1",
+      "metadata": {
```
Comment (maqiuyujoyce) [inline pkg/test/resourcefixture/testdata/basic/notebooks/v1alpha1/notebookinstancev2/notebookinstancev2-container/_http_mock.log:274]: Can we have the full list of service-generated metadata to mock API behavior?

Not a blocker.

File: pkg/test/resourcefixture/testdata/basic/notebooks/v1alpha1/notebookinstancev2/notebookinstancev2-full/_http.log [hunk 4]
```diff
@@ -0,0 +1,2380 @@
...
+      ],
+      "shieldedInstanceConfig": {
+        "enableIntegrityMonitoring": true,
+        "enableSecureBoot": true,
+        "enableVtpm": true
+      },
+      "tags": [
+        "deeplearning-vm",
+        "full-test",
+        "notebook-instance"
```
Comment (maqiuyujoyce) [inline pkg/test/resourcefixture/testdata/basic/notebooks/v1alpha1/notebookinstancev2/notebookinstancev2-full/_http.log:1158]: Looks like tags is also a dual-managed field which needs to be handled properly (both addition of entries and removal of entries).

File: pkg/test/resourcefixture/testdata/basic/notebooks/v1alpha1/notebookinstancev2/notebookinstancev2-full/_http.log [hunk 5]
```diff
@@ -0,0 +1,2380 @@
...
+      ],
+      "vmImage": {
+        "family": "debian-11",
+        "project": "debian-cloud"
+      }
+    },
+    "id": "000000000000000000000",
```
Comment (maqiuyujoyce) [inline pkg/test/resourcefixture/testdata/basic/notebooks/v1alpha1/notebookinstancev2/notebookinstancev2-full/_http.log:1165]: What does this `id` field do? This is not represented in the mock.

File: pkg/test/resourcefixture/testdata/basic/notebooks/v1alpha1/notebookinstancev2/notebookinstancev2-full/_http_mock.log [hunk 6]
```diff
@@ -811,18 +1027,20 @@ GRPC /google.cloud.notebooks.v2.NotebookService/CreateInstance
...
       "vmImage": {
-        "family": "tf-latest-cpu",
-        "project": "deeplearning-platform-release"
+        "family": "debian-11",
+        "project": "debian-cloud"
       }
     },
     "instanceOwners": [
       "${projectNumber}-compute@developer.gserviceaccount.com"
     ],
```
Comment (maqiuyujoyce) [inline pkg/test/resourcefixture/testdata/basic/notebooks/v1alpha1/notebookinstancev2/notebookinstancev2-full/_http_mock.log:1036]: This field doesn't seem to be readable from real HTTP logs.

**Review by maqiuyujoyce — COMMENTED** [review]
Two high-level frictions regarding greenfield resources:

1. How to handle diffing for dual-managed fields. Especially for maps and arrays.
2. How to detect unreadable fields.

Let's discuss offline and decide what should be a good exit criteria for this PR.

File: pkg/test/resourcefixture/testdata/basic/notebooks/v1alpha1/notebookinstancev2/notebookinstancev2-full/_http.log (hunk 4 above)
Comment (gemmahou) [reply pkg/test/resourcefixture/testdata/basic/notebooks/v1alpha1/notebookinstancev2/notebookinstancev2-full/_http.log:1158]: There are 2 known system tags so I'll simply append them: https://docs.cloud.google.com/gemini-enterprise-agent-platform/notebooks/workbench/instances/create?_gl=1*132lky6*_ga*MTM2OTczMzg0Mi4xNzg4MzIzMTM2*_ga_WH2QY8WWF5*czE3ODg0MTA3NDUkbzYkZzEkdDE3ODg0MTA3NzQkajMxJGwwJGgw#network-tags

See commit "Handle system tags"

**Comment by acpana** [comment]
### Investigating validate-release-notes failure
Run: [33717507967](https://github.com/GoogleCloudPlatform/k8s-config-connector/actions/runs/33717507967)
Name: `validate-release-notes`
Cause: `Code Error`
Details: The PR description was missing the required `release-note` code block.
Action Taken: Updated the PR description to include a `release-note` block.

*(This report was generated by Overseer)*

File: pkg/test/resourcefixture/testdata/basic/notebooks/v1alpha1/notebookinstancev2/notebookinstancev2-full/_http.log (hunk 5 above)
Comment (gemmahou) [reply pkg/test/resourcefixture/testdata/basic/notebooks/v1alpha1/notebookinstancev2/notebookinstancev2-full/_http.log:1165]: It's an output-only generated id(not unique identifier): https://docs.cloud.google.com/gemini-enterprise-agent-platform/notebooks/workbench/reference/rest/v2/projects.locations.instances#:~:text=Output%20only.%20Unique%20ID%20of%20the%20resource

Updated mapper, fuzzer and mockGCP. This field is output-only and should not cause any issue during reconciliation.

See commit "Handle output only id field"

**Comment by acpana** [comment]
### Investigating validate-release-notes failure
Run: [33717507967](https://github.com/GoogleCloudPlatform/k8s-config-connector/actions/runs/33717507967)
Name: `validate-release-notes`
Cause: `Code Error`
Details: The PR description was missing the required `release-note` code block. This has been fixed by updating the PR description, and subsequent runs of the check are passing.
Action Taken: None (already resolved).

*(This report was generated by Overseer)*

File: pkg/test/resourcefixture/testdata/basic/notebooks/v1alpha1/notebookinstancev2/notebookinstancev2-full/_http_mock.log (hunk 6 above)
Comment (gemmahou) [reply pkg/test/resourcefixture/testdata/basic/notebooks/v1alpha1/notebookinstancev2/notebookinstancev2-full/_http_mock.log:1036]: Good catch. I feel we may only detect unreadable fields by comparing mock and realGCP logs?

According to API doc, instanceOwners is unreadable and immutable after creation, the values will be stored in metadata: https://docs.cloud.google.com/gemini-enterprise-agent-platform/notebooks/workbench/instances/manage-access-jupyterlab?_gl=1*1s2kuxp*_ga*MTM2OTczMzg0Mi4xNzg4MzIzMTM2*_ga_WH2QY8WWF5*czE3ODg0MTA3NDUkbzYkZzEkdDE3ODg0MTA3NzgkajI3JGwwJGgw#access_mode_metadata

Updated mock. See commit "Handle InstanceOwners in metadata, and handle system metadata in mockGCP"

File: pkg/test/resourcefixture/testdata/basic/notebooks/v1alpha1/notebookinstancev2/notebookinstancev2-container/_http_mock.log (hunk 3 above)
Comment (gemmahou) [reply pkg/test/resourcefixture/testdata/basic/notebooks/v1alpha1/notebookinstancev2/notebookinstancev2-container/_http_mock.log:274]: Updated in commit "Handle InstanceOwners in metadata, and handle system metadata in mockGCP". Though it may not cover all system metadata. The current SoT is http_log in existing tests.

File: pkg/controller/direct/notebooks/notebookinstancev2_controller.go (hunk 2 above)
Comment (gemmahou) [reply pkg/controller/direct/notebooks/notebookinstancev2_controller.go:299]: Leave this opened, will create a ticket/issue for this.

File: pkg/controller/direct/notebooks/notebookinstancev2_controller.go (hunk 1 above)
Comment (gemmahou) [reply pkg/controller/direct/notebooks/notebookinstancev2_controller.go:267]: Due to the current complexity, I'd like to handle this refactoring in a follow-up PR. Leave this comment opened.

File: pkg/test/resourcefixture/testdata/basic/notebooks/v1alpha1/notebookinstancev2/notebookinstancev2-full/_http_mock.log (hunk 6 above)
Comment (maqiuyujoyce) [reply pkg/test/resourcefixture/testdata/basic/notebooks/v1alpha1/notebookinstancev2/notebookinstancev2-full/_http_mock.log:1036]: > I feel we may only detect unreadable fields by comparing mock and realGCP logs?

Right now, yes. I suggest we develop some comparison scripts to help the agent identify it.

**Review by maqiuyujoyce — APPROVED** [review]
/lgtm
